README.md
# flutter_culqi

A [Flutter Package]() to allow the integration of Android/iOS apps with the [Culqi API](https://www.culqi.com/docs/#/).

## How to install it?

Add the package to pubspec.yaml:
```dart
flutter_culqi : ^0.5.2
```

Import and use it! :)
``` dart
import 'package:culqi_flutter/culqi_flutter.dart';
```

## What **flutter_culqi** do?

This package comes with two main components:
1. An easy way to request the token from Culqi
2. A widget to make faster the integration

### 1. An easy way to request the token from Culqi:

**flutter_culqi** provides you with a simple interface to request and get the token from Culqi. It's easy as:

+ Create an String variable called `_publicKey`, assign to it the public key obtained from your panel at Culqi.
    ```dart
    final String _publicKey = 'INSERT_YOUR_PUBLIC_KEY_HERE';
    ```
+ Create a `CulqiCard` object (all fields are required).
    ```dart
    CulqiCard card = CulqiCard(
        cardNumber: '4111111111111111',
        cvv: '123',
        expirationMonth: 9,
        expirationYear: 2020,
        email:'test@testmail.com'
    );
    ```
+ Create a `CulqiTokenizer` object. This will keep the `CulqiCard` created above.
    ```dart
    CulqiTokenizer tokenizer = CulqiTokenizer(card);
    ```
+ And finally request the `CulqiToken`. The `_publicKey` is the same we setted above, in the first step.
    ```dart
    var result = await tokenizer.getToken(publicKey: _publicKey);
    if(result is CulqiToken){
      print(result.token);
    }else if(result is CulqiError){
      print(result);
    }
    ```
All the code can be found at the [example folder](https://github.com/limonadev/flutter_culqi/tree/master/example).

### 2. A widget to make faster the integration:

We may not want to implement the UI to request all the data (card number, expiration date, etc). Because of that, this packages comes with a widget (called `CulqiPayment`) that makes the job for us. So, how to use it?

+ The widget can be created like any other. As body of `Scaffold` for example.
    ```dart

    class TestWidget extends StatefulWidget {
        @override
        TestWidgetState createState() => TestWidgetState();
    }

    class TestWidgetState extends State<TestWidget>{
        // We need this key to get the fields inside the widget
        GlobalKey<CulqiPaymentState> _widgetKey = GlobalKey();
        
        // This is again the public key from Culqi
        final String _publicKey = 'INSERT_YOUR_PUBLIC_KEY_HERE';

        // ... More code

        @override
        Widget build(BuildContext context){
            return Scaffold(
                body: ListView(
                    children: <Widget>[
                        // Here we create it and pass the _widgetKey
                        CulqiPayment(_widgetKey),
                        RaisedButton(
                            // This function will get the token
                            // with the fields from CulqiPayment
                            onPressed: getToken,
                            child: Text('Get Token')
                        )
                    ]
                )
            );
        }
    }
    ```
+ To make clearer how to get the `CulqiToken` when using the `CulqiPayment` widget, here is the `getToken()` function called from the RaisedButton above.
    ```dart
    class TestWidgetState extends State<TestWidget>{
        // Here comes the above code
        // ...
        // Here ends the above code

        void getToken() async{
            CulqiCard _card = CulqiCard();

            // Here we are getting all the fields from the widget
            // through setInfoOn(CulqiCard _card) function
            bool success = _widgetKey.currentState.setInfoOn(_card);

            // The widget doesn't ask for email, because 
            // assumes the app already has the user email  
            // (from login for example). 
            // But in case is needed, in the example
            // folder at repository, can be found the same 
            // code but requesting the email.
            //
            // Here, for simplicity, we set the email

            _card.email = 'test@testemail.com';

            CulqiTokenizer _tokenizer = CulqiTokenizer(_card);

            var r = await tokenizer.getToken(publicKey:_publicKey);
            
            if(r is CulqiToken){
                print(r.token);
            }else if(r is CulqiError){
                print(r);
            }
        }
    }
    ```
Again, all the code can be found at the [example folder](https://github.com/limonadev/flutter_culqi/tree/master/example).

#### The `CulqiPayment` widget:

This widgets allows two optional parameters: `locale` and `years`.
+ `locale` is an String with value 'es' or 'en', meaning Spanish and English respectively. This value is used to show the widget in the corresponding language. If not provided or not supported language is provided, the default goes to 'es'.
+ `years` is a `List<int>`. Represents the year range to choose at expiration year. Default goes to [2020, 2021, 2022, 2023, 2024, 2025].

## Suggestions and bugs

Please file feature requests, suggestions and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/limonadev/flutter_culqi/issues

CHANGELOG.md
## 0.5.2

- Added a better error handling (with more specific errors from Culqi).
- Added some tests to check the specific errors from Culqi

## 0.5.1

- Fixed some maintenance suggestions from pub.dev

## 0.5.0

- Initial version, created by Limonadev.
- Support for create token from Culqi
- Provides the CulqiPayment widget

example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:flutter_culqi/flutter_culqi.dart';


/// You need to paste your Culqi public key here
final String _publicKey = 'INSERT_YOUR_PUBLIC_KEY_HERE';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {

  /// This functions tests the creation of a [CulqiToken] given a [CulqiCard]
  void onlyApi() async{
    CulqiCard card = CulqiCard(
        cardNumber: '4111111111111111',
        cvv: '123',
        expirationMonth: 9,
        expirationYear: 2020,
        email:'test@testmail.com'
    );

    CulqiTokenizer tokenizer = CulqiTokenizer(card);

    var result = await tokenizer.getToken(publicKey: _publicKey);
    if(result is CulqiToken){
      print(result.token);
    }else if(result is CulqiError){
      print(result);
    }

  }

  MyApp(){
    onlyApi();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Culqi Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {

  @override
  _MyHomePageState createState() => _MyHomePageState();

}

class _MyHomePageState extends State<MyHomePage> {

  GlobalKey<CulqiPaymentState> _key = GlobalKey();

  String _email;

  String _token = '';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Culqi Example'),
      ),
      body: ListView(
        children: <Widget>[
          /// This is a provided widget to make faster the Culqi integration. Is not mandatory to use it,
          /// you can create your own form to get the necessary fields to get a token.
          ///
          /// Of course, the creation of the token requires an email, but on the most cases, the email
          /// is the login method, so the app already has the user email. Because of that the
          /// [CulqiPayment] widget wont request the email, and you need to provide it to the [CulqiCard]
          CulqiPayment(_key, locale: 'en', years: [2020, 2021, 2022, 2023, 2024],),

          /// In this example, here we are requesting the email and save it on [_email].
          ListTile(
            title: Text('Email', style: TextStyle(fontSize: 12),),
            subtitle: TextField(
              onChanged: (value){
                _email = value;
              },
              decoration: InputDecoration(
                hintText: 'test@test.com'
              ),
            ),
          ),
          Container(
            alignment: Alignment.centerRight,
            child: RaisedButton(
              onPressed: getToken,
              child: Text('Get Token', style: TextStyle(color: Colors.white),),
              color: Colors.blue,
            ),
          ),
          Text(_token)
        ],
      ),
    );
  }


  /// This async function allows us to get the info from the [CulqiPayment] widget and check if all the fields
  /// were completed. If it happens, set the [_email] on the [_card].
  ///
  /// Finally, using the [CulqiTokenizer] we try to get the token.
  void getToken() async{
    CulqiCard _card = CulqiCard();
    bool success = _key.currentState.setInfoOn(_card);

    if(success){
      _card.email = _email;

      CulqiTokenizer _tokenizer = CulqiTokenizer(_card);

      var result = await _tokenizer.getToken(publicKey: _publicKey);
      if(result is CulqiToken){
        setState(() {
          _token = result.token;
        });
      }else if(result is CulqiError){
        setState(() {
          _token = result.errorMessage;
        });
      }
    }
  }
}
